src/libmongoc/src/mongoc/mongoc-scram-private.h

```diff
@@ -123,6 +123,346 @@ _mongoc_sasl_prep_required (const char *str);
 char *
 _mongoc_sasl_prep (const char *in_utf8, int in_utf8_len, bson_error_t *err);
 
+/* returns how many bytes a UTF-8 character is. */
+int
+_mongoc_utf8_char_length (const char *c);
+
+/* returns how many characters are in a UTF-8 string. Returns -1 on error. */
+int
+_mongoc_utf8_string_length (const char *s);
+
+/* returns whether a UTF-8 character is valid or not. */
+bool
+_mongoc_utf8_is_valid (const char *c, int length);
+
+/* returns whether a character is between two limits (inclusive). */
+bool
+_mongoc_char_between_chars (const char c, const char lower, const char upper);
```

Comment:
Use `uint8_t` to represent code units, even if the UTF-8 string is represented as `char*`.